data-plane/bindings/adapter/src/adapter.rs

```diff
@@ -0,0 +1,1841 @@
+impl FfiCompletionHandle {
+    /// Wait for the operation to complete (blocking version for FFI)
+    ///
+    /// This blocks the calling thread until the operation completes.
+    /// Use this from Go or other languages when you need to ensure
+    /// an operation has finished before proceeding.
+    ///
+    /// **Note:** This can only be called once per handle. Subsequent calls
+    /// will return an error.
+    ///
+    /// # Returns
+    /// * `Ok(())` - Operation completed successfully
+    /// * `Err(SlimError)` - Operation failed or handle already consumed
+    pub fn wait(&self) -> Result<(), SlimError> {
+        let receiver = self
+            .receiver
+            .lock()
+            .take()
+            .ok_or_else(|| SlimError::InternalError {
+                message: "CompletionHandle already consumed (wait can only be called once)"
+                    .to_string(),
+            })?;
+
+        self.runtime.block_on(async {
+            receiver
+                .await
+                .map_err(|_| SlimError::InternalError {
+                    message: "Completion sender dropped before result was sent".to_string(),
+                })?
+                .map_err(|e| SlimError::SessionError {
+                    message: e.to_string(),
+                })
+        })
+    }
+
+    /// Wait for the operation to complete (async version)
+    ///
+    /// This is the async version that integrates with UniFFI's polling mechanism.
+    /// The operation will yield control while waiting.
+    ///
+    /// **Note:** This can only be called once per handle. Subsequent calls
+    /// will return an error.
+    ///
+    /// # Returns
+    /// * `Ok(())` - Operation completed successfully
+    /// * `Err(SlimError)` - Operation failed or handle already consumed
+    pub async fn wait_async(&self) -> Result<(), SlimError> {
+        let receiver = self
+            .receiver
+            .lock()
+            .take()
+            .ok_or_else(|| SlimError::InternalError {
+                message: "CompletionHandle already consumed (wait can only be called once)"
+                    .to_string(),
+            })?;
+
+        receiver
+            .await
+            .map_err(|_| SlimError::InternalError {
+                message: "Completion sender dropped before result was sent".to_string(),
+            })?
+            .map_err(|e| SlimError::SessionError {
+                message: e.to_string(),
+            })
+    }
+}
+
+// ============================================================================
+// FFI SessionContext Wrapper
+// ============================================================================
+
+/// FFISessionContext represents an active session (FFI-compatible wrapper)
+#[derive(uniffi::Object)]
+pub struct FFISessionContext {
+    /// The inner BindingsSessionContext (public for Python bindings access)
+    pub inner: crate::BindingsSessionContext,
+    runtime: Arc<tokio::runtime::Runtime>,
```

Comment:
Given that we don't need to hold a reference to the runtime, do we still have a reason to duplicate the SessionContext here?